### PR TITLE
abfs: support blob storage versioning

### PR DESCRIPTION
**Pull request by pmrowla**
Implements `AzureBlobFileSystem` support for [blob versioning](https://docs.microsoft.com/en-us/azure/storage/blobs/versioning-overview) equivalent to what is currently supported in s3fs

- Filesystem must be configured with `version_aware=True` (and container must have versioning enabled)
- `version_id` property returned in fs read methods that return file details (i.e. `find`, `ls`, `info`)
    - Defaults to `None` when filesystem or container is not versioning-aware
- For file read methods specific blob version ID can be specified via `version_id` kwarg or via the requested path using `versionid` query parameter
    - path example: `abfs://my_container/path/to/versioned_file?versionid=some_version_id`
    - `versionid` used in path instead of `version_id` so paths are consistent with the Azure REST API [URI format](https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob#uri-parameters). This is consistent with the s3fs implementation which uses `version_id` in the fsspec python API and `VersionId` in paths (to match the AWS URI format)

Regarding testing:
- Version-aware filesystem methods have been manually tested against actual Azure containers with versioning enabled
- Azurite does not support blob versioning (it will ignore any versioning related API arguments)
- Version-aware filesystem methods are currently tested with mocked Azure SDK calls to ensure that we at least make calls with the correct arguments, and can parse/filter any returned version-aware blob properties

~~TODO:~~
- [x] figure out easiest test implementation (azurite does not currently implement any of the blob versioning functionality: https://github.com/Azure/Azurite/issues/665)

### CODE REVIEW

File: adlfs/tests/conftest.py
Comment (pmrowla) [inline adlfs/tests/conftest.py:62]: `get_event_loop()` is deprecated in 3.10, and separate sessions should use their own loops

File: adlfs/tests/conftest.py
Comment (pmrowla) [inline adlfs/tests/conftest.py:68]: with pytest-asyncio`mode=auto` we need to explicitly update the policy eventloop after closing otherwise `asyncio` may overwrite it: https://github.com/pytest-dev/pytest-asyncio/issues/257

**Review by efiop — APPROVED** [review]
LGTM, but I have a conflict of interest here 😄 (we need it for active use in dvc) Waiting for another review.

File: setup.py
Comment (pmrowla) [inline setup.py:36]: Blob versioning related adlfs changes require using the current azure service version and pagination changes from 12.12.0:

https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/storage/azure-storage-blob/CHANGELOG.md#12120b1-2022-04-14

**Comment by efiop** [comment]
@TomAugspurger @hayesgb Hey guys. 🙂  Could you please take a look at this PR when you have time? 🙏

**Comment by efiop** [comment]
It feels a bit intrusive from my side, but it looks like there are no comments (I'm sure people are just busy with other stuff), old and new tests pass and we are ready to help maintain this, so I hope no one will mind me taking the liberty to move forward with this 🙂

**Comment by efiop** [comment]
@TomAugspurger Thank you! 🙏

File: adlfs/utils.py
Comment (efiop) [inline adlfs/utils.py:6]: version_id could actually be present but set to `None`, which will make us ignore blobs when `fs.version_aware==True`. So maybe

``suggestion
        version_id is None and (blob.get("version_id") is None or blob["is_current_version"])
```

?

Comment (efiop) [reply adlfs/utils.py:6]: Also, there is `blob["version_id"]` a line below, maybe we shouldn't assume that it is always present in `blob` too and do a `get()` instead?

Comment (pmrowla) [reply adlfs/utils.py]: should be fixed in bbd9483

**Comment by hayesgb** [comment]
Thanks.  This looks great!

**Pull request merged**
